beacon-chain/sync/rpc_blob_sidecars_by_range.go

```diff
@@ -0,0 +1,178 @@
+	ctx, span := trace.StartSpan(ctx, "sync.BlobsSidecarsByRangeHandler")
+	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, respTimeout)
+	defer cancel()
+	SetRPCStreamDeadlines(stream)
+	log := log.WithField("handler", p2p.BlobSidecarsByRangeName[1:]) // slice the leading slash off the name var
+
+	r, ok := msg.(*pb.BlobSidecarsByRangeRequest)
+	if !ok {
+		return errors.New("message is not type *pb.BlobsSidecarsByRangeRequest")
+	}
+	if err := s.rateLimiter.validateRequest(stream, 1); err != nil {
+		return err
+	}
+	rp, err := validateBlobsByRange(r, s.cfg.chain.CurrentSlot())
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	// Ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	batcher, err := newBlockRangeBatcher(rp, s.cfg.beaconDB, s.rateLimiter, s.cfg.chain.IsCanonical, ticker)
+
+	var batch blockBatch
+	var totalWrites uint64
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
+		batchStart := time.Now()
+		writes, err := s.streamBlobBatch(ctx, batch, stream)
+		rpcBlobsByRangeResponseLatency.Observe(float64(time.Since(batchStart).Milliseconds()))
+		if err != nil {
+			return err
+		}
+		totalWrites += writes
+		// once we have written MAX_REQUEST_BLOB_SIDECARS, we're done serving the request
+		if totalWrites >= params.BeaconNetworkConfig().MaxRequestBlobsSidecars {
+			break
+		}
+	}
+	if err := batch.error(); err != nil {
+		log.WithError(err).Debug("error in BlocksByRange batch")
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	closeStream(stream, log)
+	return nil
+}
+
+func blobsByRangeMinStartSlot(current primitives.Slot) (primitives.Slot, error) {
+	minReqEpochs := params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
+	currEpoch := slots.ToEpoch(current)
+	minStart := params.BeaconConfig().DenebForkEpoch
+	if currEpoch > minReqEpochs && currEpoch-minReqEpochs > minStart {
+		minStart = currEpoch - minReqEpochs
+	}
+	return slots.EpochStart(minStart)
+}
+
+func blobBatchLimit() uint64 {
+	return uint64(flags.Get().BlockBatchLimit / fieldparams.MaxBlobsPerBlock)
+}
+
+func validateBlobsByRange(r *pb.BlobSidecarsByRangeRequest, current primitives.Slot) (rangeParams, error) {
+	rp := rangeParams{
+		start: r.StartSlot,
+		size:  r.Count,
+	}
+	// Peers may overshoot the current slot when in initial sync, so we don't want to penalize them by treating the
+	// request as an error. So instead we return a set of params that acts as a noop.
+	if rp.start > current {
+		return rangeParams{start: current, end: current, size: 0}, nil
+	}
+
+	var err error
+	rp.end, err = rp.start.SafeAdd((rp.size - 1))
```

Comment:
underflow if `rp.size` is 0?

Reply:
added this validation: https://github.com/prysmaticlabs/prysm/pull/12499/commits/e84e08aea8c847f655a583476a6d7fb3a8c2f679